Mathlib/Tactic/HashCommandLinter.lean

```diff
@@ -0,0 +1,75 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Lean.Elab.Command
+import Lean.Linter.Util
+
+/-!
+#  `#`-command linter
+
+The `#`-command linter produces a warning when a command starting with `#` is used *and*
+the command emits no message.
+-/
+
+namespace Mathlib.Linter
+
+/--
+The linter emits a warning on any command beginning with `#` that itself emits no message.
+For example, `#guard true` and `#check_tactic True ~> True by skip` trigger a message.
+There is a `whitelist` of silent `#`-command that are allowed.
+-/
+register_option linter.hashCommand : Bool := {
+  defValue := true
+  descr := "enable the `#`-command linter"
+}
+
+namespace HashCommandLinter
+
+open Lean Elab
+
+/-- Gets the value of the `linter.hashCommand` option. -/
+def getLinterHash (o : Options) : Bool := Linter.getLinterValue linter.hashCommand o
+
+open Command in
+/-- Exactly like `withSetOptionIn`, but recursively discards nested uses of `in`.
+Intended to be used in the `hashCommand` linter, where we want to enter `set_option` `in` commands.
+-/
+private partial def withSetOptionIn' (cmd : CommandElab) : CommandElab := fun stx => do
+  if stx.getKind == ``Lean.Parser.Command.in then
+    if stx[0].getKind == ``Lean.Parser.Command.set_option then
+      let opts ← Elab.elabSetOption stx[0][1] stx[0][2]
+      Command.withScope (fun scope => { scope with opts }) do
+        withSetOptionIn' cmd stx[2]
+    else
+      withSetOptionIn' cmd stx[2]
+  else
+    cmd stx
+
+/-- `whitelist` is the array of `#`-commands that are allowed in 'Mathlib'. -/
+private abbrev whitelist : HashSet String := HashSet.empty
+  |>.insert "#align"
+  |>.insert "#align_import"
+  |>.insert "#noalign"
```

Comment:
I hope `Std.Lean.HashSet` gets upstreamed, because then you could write `{ "#align", "#align_import", "#noalign" }` without adding an import! :) Up to you if you want to prioritize fewer imports or shorter syntax, just wanted to mention it.

Reply:
I like the notation and was not aware of it!  I added the import, since this linter is going into Mathlib, I take a nonchalant stance to "assuming" `Core` and `Std` files.